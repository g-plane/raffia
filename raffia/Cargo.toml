[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
