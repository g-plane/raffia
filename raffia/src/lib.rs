//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
