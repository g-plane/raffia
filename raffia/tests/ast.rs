use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
