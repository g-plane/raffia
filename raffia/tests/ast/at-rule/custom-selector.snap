---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "custom-selector",
        raw: "custom-selector",
        span: Span(
          start: 1,
          end: 16,
        ),
      ),
      prelude: Some(CustomSelectorPrelude(
        type: "CustomSelectorPrelude",
        customSelector: CustomSelector(
          type: "CustomSelector",
          prefixArg: None,
          name: Ident(
            type: "Ident",
            name: "--foo",
            raw: "--foo",
            span: Span(
              start: 18,
              end: 23,
            ),
          ),
          args: None,
          span: Span(
            start: 18,
            end: 23,
          ),
        ),
        selector: SelectorList(
          type: "SelectorList",
          selectors: [
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "foo",
                        raw: "foo",
                        span: Span(
                          start: 25,
                          end: 28,
                        ),
                      ),
                      span: Span(
                        start: 24,
                        end: 28,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 24,
                    end: 28,
                  ),
                ),
              ],
              span: Span(
                start: 24,
                end: 28,
              ),
            ),
          ],
          commaSpans: [],
          span: Span(
            start: 24,
            end: 28,
          ),
        ),
        span: Span(
          start: 18,
          end: 28,
        ),
      )),
      block: None,
      span: Span(
        start: 0,
        end: 28,
      ),
    ),
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "custom-selector",
        raw: "custom-selector",
        span: Span(
          start: 31,
          end: 46,
        ),
      ),
      prelude: Some(CustomSelectorPrelude(
        type: "CustomSelectorPrelude",
        customSelector: CustomSelector(
          type: "CustomSelector",
          prefixArg: None,
          name: Ident(
            type: "Ident",
            name: "--bar",
            raw: "--bar",
            span: Span(
              start: 48,
              end: 53,
            ),
          ),
          args: None,
          span: Span(
            start: 48,
            end: 53,
          ),
        ),
        selector: SelectorList(
          type: "SelectorList",
          selectors: [
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "bar",
                        raw: "bar",
                        span: Span(
                          start: 55,
                          end: 58,
                        ),
                      ),
                      span: Span(
                        start: 54,
                        end: 58,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 54,
                    end: 58,
                  ),
                ),
              ],
              span: Span(
                start: 54,
                end: 58,
              ),
            ),
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "baz",
                        raw: "baz",
                        span: Span(
                          start: 61,
                          end: 64,
                        ),
                      ),
                      span: Span(
                        start: 60,
                        end: 64,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 60,
                    end: 64,
                  ),
                ),
              ],
              span: Span(
                start: 60,
                end: 64,
              ),
            ),
          ],
          commaSpans: [
            Span(
              start: 58,
              end: 59,
            ),
          ],
          span: Span(
            start: 54,
            end: 64,
          ),
        ),
        span: Span(
          start: 48,
          end: 64,
        ),
      )),
      block: None,
      span: Span(
        start: 30,
        end: 64,
      ),
    ),
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "custom-selector",
        raw: "custom-selector",
        span: Span(
          start: 67,
          end: 82,
        ),
      ),
      prelude: Some(CustomSelectorPrelude(
        type: "CustomSelectorPrelude",
        customSelector: CustomSelector(
          type: "CustomSelector",
          prefixArg: Some(CustomSelectorArg(
            type: "CustomSelectorArg",
            name: Ident(
              type: "Ident",
              name: "foo",
              raw: "foo",
              span: Span(
                start: 84,
                end: 87,
              ),
            ),
            span: Span(
              start: 83,
              end: 87,
            ),
          )),
          name: Ident(
            type: "Ident",
            name: "--foo",
            raw: "--foo",
            span: Span(
              start: 88,
              end: 93,
            ),
          ),
          args: None,
          span: Span(
            start: 83,
            end: 93,
          ),
        ),
        selector: SelectorList(
          type: "SelectorList",
          selectors: [
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "foo",
                        raw: "foo",
                        span: Span(
                          start: 95,
                          end: 98,
                        ),
                      ),
                      span: Span(
                        start: 94,
                        end: 98,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 94,
                    end: 98,
                  ),
                ),
              ],
              span: Span(
                start: 94,
                end: 98,
              ),
            ),
          ],
          commaSpans: [],
          span: Span(
            start: 94,
            end: 98,
          ),
        ),
        span: Span(
          start: 83,
          end: 98,
        ),
      )),
      block: None,
      span: Span(
        start: 66,
        end: 98,
      ),
    ),
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "custom-selector",
        raw: "custom-selector",
        span: Span(
          start: 101,
          end: 116,
        ),
      ),
      prelude: Some(CustomSelectorPrelude(
        type: "CustomSelectorPrelude",
        customSelector: CustomSelector(
          type: "CustomSelector",
          prefixArg: Some(CustomSelectorArg(
            type: "CustomSelectorArg",
            name: Ident(
              type: "Ident",
              name: "bar",
              raw: "bar",
              span: Span(
                start: 118,
                end: 121,
              ),
            ),
            span: Span(
              start: 117,
              end: 121,
            ),
          )),
          name: Ident(
            type: "Ident",
            name: "--bar",
            raw: "--bar",
            span: Span(
              start: 122,
              end: 127,
            ),
          ),
          args: Some(CustomSelectorArgs(
            type: "CustomSelectorArgs",
            args: [],
            commaSpans: [],
            span: Span(
              start: 127,
              end: 129,
            ),
          )),
          span: Span(
            start: 117,
            end: 129,
          ),
        ),
        selector: SelectorList(
          type: "SelectorList",
          selectors: [
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "bar",
                        raw: "bar",
                        span: Span(
                          start: 131,
                          end: 134,
                        ),
                      ),
                      span: Span(
                        start: 130,
                        end: 134,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 130,
                    end: 134,
                  ),
                ),
              ],
              span: Span(
                start: 130,
                end: 134,
              ),
            ),
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "baz",
                        raw: "baz",
                        span: Span(
                          start: 137,
                          end: 140,
                        ),
                      ),
                      span: Span(
                        start: 136,
                        end: 140,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 136,
                    end: 140,
                  ),
                ),
              ],
              span: Span(
                start: 136,
                end: 140,
              ),
            ),
          ],
          commaSpans: [
            Span(
              start: 134,
              end: 135,
            ),
          ],
          span: Span(
            start: 130,
            end: 140,
          ),
        ),
        span: Span(
          start: 117,
          end: 140,
        ),
      )),
      block: None,
      span: Span(
        start: 100,
        end: 140,
      ),
    ),
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "custom-selector",
        raw: "custom-selector",
        span: Span(
          start: 143,
          end: 158,
        ),
      ),
      prelude: Some(CustomSelectorPrelude(
        type: "CustomSelectorPrelude",
        customSelector: CustomSelector(
          type: "CustomSelector",
          prefixArg: Some(CustomSelectorArg(
            type: "CustomSelectorArg",
            name: Ident(
              type: "Ident",
              name: "bar",
              raw: "bar",
              span: Span(
                start: 160,
                end: 163,
              ),
            ),
            span: Span(
              start: 159,
              end: 163,
            ),
          )),
          name: Ident(
            type: "Ident",
            name: "--bar",
            raw: "--bar",
            span: Span(
              start: 164,
              end: 169,
            ),
          ),
          args: Some(CustomSelectorArgs(
            type: "CustomSelectorArgs",
            args: [
              CustomSelectorArg(
                type: "CustomSelectorArg",
                name: Ident(
                  type: "Ident",
                  name: "a",
                  raw: "a",
                  span: Span(
                    start: 171,
                    end: 172,
                  ),
                ),
                span: Span(
                  start: 170,
                  end: 172,
                ),
              ),
              CustomSelectorArg(
                type: "CustomSelectorArg",
                name: Ident(
                  type: "Ident",
                  name: "b",
                  raw: "b",
                  span: Span(
                    start: 175,
                    end: 176,
                  ),
                ),
                span: Span(
                  start: 174,
                  end: 176,
                ),
              ),
            ],
            commaSpans: [
              Span(
                start: 172,
                end: 173,
              ),
            ],
            span: Span(
              start: 169,
              end: 177,
            ),
          )),
          span: Span(
            start: 159,
            end: 177,
          ),
        ),
        selector: SelectorList(
          type: "SelectorList",
          selectors: [
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "bar",
                        raw: "bar",
                        span: Span(
                          start: 179,
                          end: 182,
                        ),
                      ),
                      span: Span(
                        start: 178,
                        end: 182,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 178,
                    end: 182,
                  ),
                ),
              ],
              span: Span(
                start: 178,
                end: 182,
              ),
            ),
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "baz",
                        raw: "baz",
                        span: Span(
                          start: 185,
                          end: 188,
                        ),
                      ),
                      span: Span(
                        start: 184,
                        end: 188,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 184,
                    end: 188,
                  ),
                ),
              ],
              span: Span(
                start: 184,
                end: 188,
              ),
            ),
          ],
          commaSpans: [
            Span(
              start: 182,
              end: 183,
            ),
          ],
          span: Span(
            start: 178,
            end: 188,
          ),
        ),
        span: Span(
          start: 159,
          end: 188,
        ),
      )),
      block: None,
      span: Span(
        start: 142,
        end: 188,
      ),
    ),
    AtRule(
      type: "AtRule",
      name: Ident(
        type: "Ident",
        name: "custom-selector",
        raw: "custom-selector",
        span: Span(
          start: 191,
          end: 206,
        ),
      ),
      prelude: Some(CustomSelectorPrelude(
        type: "CustomSelectorPrelude",
        customSelector: CustomSelector(
          type: "CustomSelector",
          prefixArg: None,
          name: Ident(
            type: "Ident",
            name: "--baz",
            raw: "--baz",
            span: Span(
              start: 208,
              end: 213,
            ),
          ),
          args: Some(CustomSelectorArgs(
            type: "CustomSelectorArgs",
            args: [
              CustomSelectorArg(
                type: "CustomSelectorArg",
                name: Ident(
                  type: "Ident",
                  name: "a",
                  raw: "a",
                  span: Span(
                    start: 215,
                    end: 216,
                  ),
                ),
                span: Span(
                  start: 214,
                  end: 216,
                ),
              ),
              CustomSelectorArg(
                type: "CustomSelectorArg",
                name: Ident(
                  type: "Ident",
                  name: "b",
                  raw: "b",
                  span: Span(
                    start: 219,
                    end: 220,
                  ),
                ),
                span: Span(
                  start: 218,
                  end: 220,
                ),
              ),
              CustomSelectorArg(
                type: "CustomSelectorArg",
                name: Ident(
                  type: "Ident",
                  name: "c",
                  raw: "c",
                  span: Span(
                    start: 223,
                    end: 224,
                  ),
                ),
                span: Span(
                  start: 222,
                  end: 224,
                ),
              ),
            ],
            commaSpans: [
              Span(
                start: 216,
                end: 217,
              ),
              Span(
                start: 220,
                end: 221,
              ),
            ],
            span: Span(
              start: 213,
              end: 225,
            ),
          )),
          span: Span(
            start: 208,
            end: 225,
          ),
        ),
        selector: SelectorList(
          type: "SelectorList",
          selectors: [
            ComplexSelector(
              type: "ComplexSelector",
              children: [
                CompoundSelector(
                  type: "CompoundSelector",
                  children: [
                    ClassSelector(
                      type: "ClassSelector",
                      name: Ident(
                        type: "Ident",
                        name: "baz",
                        raw: "baz",
                        span: Span(
                          start: 227,
                          end: 230,
                        ),
                      ),
                      span: Span(
                        start: 226,
                        end: 230,
                      ),
                    ),
                  ],
                  span: Span(
                    start: 226,
                    end: 230,
                  ),
                ),
              ],
              span: Span(
                start: 226,
                end: 230,
              ),
            ),
          ],
          commaSpans: [],
          span: Span(
            start: 226,
            end: 230,
          ),
        ),
        span: Span(
          start: 208,
          end: 230,
        ),
      )),
      block: None,
      span: Span(
        start: 190,
        end: 230,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 232,
  ),
)
